[iOS] Add Live Text option to system context menu

jingshao-code opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

**Live Text (OCR) button disappeared from text field menus on iOS after the Secure Paste M2 update. This PR adds it back.**

**Fixes #169781**

Note: This is a draft PR for initial review. Still need to add tests or split to framework and engine

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

jingshao-code commented:
Hi @justinmc @hellohuanlin ,
Draft PR ready for initial review.
Note: CI failing on SDK analysis seems unrelated to my changes.

justinmc reviewed (commented):
The approach looks good to me! Minus some small comments and missing tests.

I was struggling to build the engine with Xcode 26, so I wasn't able to try this locally for now. I'm going to go back to 16.4 and try again.

*Edit*: I was able to get this to run locally with Xcode 16.4, and everything worked as expected 👍 .

hellohuanlin reviewed (commented):
Reviewed the engine part

justinmc commented on packages/flutter/lib/src/widgets/system_context_menu.dart:
```diff
@@ -414,5 +415,34 @@ final class IOSSystemContextMenuItemShare extends IOSSystemContextMenuItem with
+  @override
+  IOSSystemContextMenuItemData getData(WidgetsLocalizations localizations) {
+    return IOSSystemContextMenuItemDataLiveText(
+      title: 'Scan Text',
```
Ah that's right. Sounds good to me.

justinmc commented on packages/flutter/lib/src/widgets/system_context_menu.dart:
```diff
@@ -414,5 +415,34 @@ final class IOSSystemContextMenuItemShare extends IOSSystemContextMenuItem with
+/// Should only appear when the text field is not secure, has focus, and has
+/// no selection.
+///
+/// The title and action are both handled by the platform.
```
Actually per @hellohuanlin's comment above (https://github.com/flutter/flutter/pull/170969#discussion_r2162355299), this should not localize the string in the framework.

jingshao-code committed:
Fix iOS Live Text button not showing in context menu

- Fix variable naming: hasSelection -> hasNoSelection
- Remove unnecessary respondsToSelector check
- Make title parameter optional for Live Text
- Use debugFillProperties instead of toString
- Let platform handle Live Text localization

jingshao-code committed:
Merge branch 'master' into fix-ios15-ocr-menu

jingshao-code committed:
Add tests for Live Text in system context menu

justinmc commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -728,4 +728,25 @@ void main() {
+  );
+
+  test('IOSSystemContextMenuItemDataLiveText debugFillProperties', () {
+    const String? title = null;
+    const IOSSystemContextMenuItemDataLiveText item = IOSSystemContextMenuItemDataLiveText(title: title);
```
You're getting an analyzer failure here because `null` is the default value, so you don't need to pass it.

justinmc reviewed (commented):
I left a comment about the analyzer failure, and there also seems to be a formatting failure in FlutterTextInputPluginTest.mm.

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1314,6 +1322,14 @@ - (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
+  } else if (action == @selector(captureTextFromCamera:)) {
+    if (@available(iOS 15.0, *)) {
+      BOOL isSecure = self.isSecureTextEntry;
+      BOOL hasSelection = _selectedTextRange.empty;
```
Discussed offline - we can simply return YES here. The logic is already done by the framework. 

We have checks for other selectors (e.g. copy/paste) because they can be triggered by keyboard shortcut.

jingshao-code replied:
Thanks for the offline discussion! I've simplified the implementation to just return YES in commit 9c7b184.

jingshao-code committed:
Fix analyzer error and simplify implementation

- Simplify Live Text implementation in FlutterTextInputPlugin.mm
- Remove unnecessary null assignment in test

jingshao-code committed:
Fix Live Text test after simplifying canPerformAction logic

jingshao-code commented:
Fixed the failing test in bf114d6 - updated to match the simplified implementation.

jingshao-code committed:
Remove unnecessary captureTextFromCamera method

jingshao-code committed:
Merge branch 'master' into fix-ios15-ocr-menu

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:
```diff
@@ -3869,4 +3869,22 @@ - (void)testInteractiveKeyboardShowKeyboardAndRemoveScreenshotAnimationIsReenabl
   [self waitForExpectations:@[ expectation ] timeout:10.0];
 }
 
+- (void)testCanPerformActionCaptureTextFromCamera {
```
1. if you need to create a new test, please move it close to the context menu tests. 
2. Can you modify the existing test case related to context menu?

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.mm:
```diff
@@ -3869,4 +3869,22 @@ - (void)testInteractiveKeyboardShowKeyboardAndRemoveScreenshotAnimationIsReenabl
+    NSArray<FlutterTextInputView*>* inputFields = self.installedInputViews;
+    FlutterTextInputView* inputView = inputFields[0];
+
+    // Live Text is always available on iOS 15+ per framework logic
```
it isn't clear what this means to me. On framework logic, it's not always available

jingshao-code replied:
I have removed this inaccurate comment. The framework does check the conditions before enabling Live Text.

hellohuanlin commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -3062,5 +3062,32 @@ final class IOSSystemContextMenuItemDataShare extends IOSSystemContextMenuItemDa
+  const IOSSystemContextMenuItemDataLiveText({this.title});
+
+  @override
+  final String? title;
```
as discussed in my previous message, title isn't needed here. Please follow the example of copy/paste/select actions.

hellohuanlin commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -728,4 +728,23 @@ void main() {
   });
+
+  test(
+    'can get the IOSSystemContextMenuItemData representation of an IOSSystemContextMenuItemLiveText',
```
please move this test case closer to other similar tests

jingshao-code replied:
Thanks for reviewing! All comments addressed:

- Moved tests next to similar tests
- No title parameter (follows copy/paste/select pattern)

jingshao-code committed:
Live Text: drop title param, match native menu pattern, move tests near system-menu cases

jingshao-code committed:
Update Live Text comment to remove unnecessary conditions

hellohuanlin reviewed (approved):
Engine side LGTM. Please wait for @justinmc's framework side approval before landing it.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -3062,5 +3062,34 @@ final class IOSSystemContextMenuItemDataShare extends IOSSystemContextMenuItemDa
+  @override
+  String get _jsonType => 'captureTextFromCamera';
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+  }
```
I guess now that there is no title here, Diagnosticable is not needed.

justinmc commented on packages/flutter/test/widgets/system_context_menu_test.dart:
```diff
@@ -699,6 +701,28 @@ void main() {
+    },
+  );
+
+  test('IOSSystemContextMenuItemDataLiveText debugFillProperties', () {
+    const IOSSystemContextMenuItemDataLiveText item = IOSSystemContextMenuItemDataLiveText();
+    final List<DiagnosticsNode> diagnosticsNodes = item.toDiagnosticsNode().getProperties();
+    expect(diagnosticsNodes, hasLength(0));
+  });
```
Also this wouldn't be necessary anymore if you remove Diagnosticable from the class.

justinmc reviewed (commented):
Some very small changes here, otherwise looks good.

jingshao-code commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -3062,5 +3062,34 @@ final class IOSSystemContextMenuItemDataShare extends IOSSystemContextMenuItemDa
   }
 }
 
+/// A [IOSSystemContextMenuItemData] for the system's built-in Live Text
```
I noticed that all other `IOSSystemContextMenuItemData` subclasses in this file use "A" rather than "An" in their documentation. Should I keep it as "A" for consistency with the existing pattern, or would you prefer I change just this one to "An"?

jingshao-code replied:
Done! Created #172019 to fix all the grammar issues

jingshao-code committed:
Remove Diagnosticable from IOSSystemContextMenuItemDataLiveText

- Remove Diagnosticable mixin as there are no properties to debug
- Remove empty debugFillProperties method
- Remove corresponding test case

jingshao-code committed:
Fix formatting in system_context_menu_test.dart

jingshao-code committed:
Merge branch 'master' into fix-ios15-ocr-menu

justinmc reviewed (approved):
LGTM 👍
